HDMI CEC - support for devices and commands

konikvranik opened the pull request:
**Description:**

This pullrequest extends functionality of previous hdmi_cec platform so now detected HDMI devices are handled as switch entites, reports name, vendor, address and state, could be turned on/off.
Also adds services to pass arbitraty CEC command and control volume.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation:** home-assistant/home-assistant.github.io#1642

**Example entry for `configuration.yaml` (if applicable):**
```yaml
hdmi_cec:
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

Commit:
cec client object

Commit:
cec command structure

Commit:
autodetect source

Commit:
volume support and native source select

Commit:
switch device

Commit:
media player device

Commit:
detecting of state

Commit:
friendly names

Commit:
hdmi cec properties

Commit:
presence detection

Commit:
simplified callbacks

Commit:
stable names

Commit:
renamed methods

Commit:
code cleanup

Commit:
name with vendor

Commit:
fixed standby call name

Commit:
fake standby/poweron

konikvranik committed:
domain switch

konikvranik committed:
domain switch

Commit:
async updating

Commit:
update separated

Commit:
cec -> hass event bridge

Commit:
fixed name generation

Commit:
code cleanup

Commit:
code cleanup

Commit:
icon constants

Commit:
code cleanup

Commit:
do not register unavailable devices

konikvranik committed:
discovery of deevices

konikvranik committed:
code cleanup

Commit:
cec device discovery

Commit:
moved method implementation into child

Commit:
service descriptions

Commit:
service descriptions

Commit:
service descriptions

Commit:
service descriptions

Commit:
changed entity init sequence

Commit:
logging cleanup

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+    _LOGGER.debug("CEC setup")
+    config = base_config.get(DOMAIN)
+
+    global CEC_CLIENT
```
Don't use a global but use an entry in the dictionary `hass.data` instead.

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+        self._available = False
+        self._hidden = False
+        self._name = None
+        DEVICE_PRESENCE[logical] = True
```
Please use `hass.data`

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+SCAN_INTERVAL = 30
 
-MAX_DEPTH = 4
+VENDORS = {0x000039: 'Toshiba',
```
There is too much protocol specific logic in this component. This should be moved to a third party lib on which Home Assistant can depend.

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+            dev_type = 'switch'
+
+            while True:
+                new_devices = filter(lambda x: CEC_CLIENT.lib_cec.PollDevice(x),
```
Please rewrite this using `for` loops to keep this code understandable.

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+    exclude = config.get(CONF_EXCLUDE)
+    stop = threading.Event()
 
     def _start_cec(event):
```
You cannot have an event handler that never returns, you are then stealing a thread from HASS. If you want to run in a thread, start your own.

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
-            _LOGGER.error("No CEC adapter found")
-            return
+        # initialise libCEC and enter the main loop
+        if CEC_CLIENT.init_lib_cec():
```
Use a guard clause

```python
if not CEC_CLIENT.init_lib_cec():
    return False

…
```

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+
+
+def icon_by_type(cec_type):
+    return ICON_TV if cec_type == 0 \
```
Code would be more readable if you use a dict: `ICONS.get(cec_type, ICON_UNKNOWN)`

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+        call.data[addr]
+
+
+class CecClient:
```
This should be moved into third party lib.

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+
+    @asyncio.coroutine
+    def async_update_availability(self):
+        self._available = self.cec_client.lib_cec.PollDevice(self._logical_address)
```
You should not do I/O inside coroutines.

kellerza replied:
If you have a coroutine, all IO should also be handled via a coroutine, so if `PollDevice` was a coroutine, you could have done:

```
_available = yield PollDevice()
```

There are ways to schedule non coroutine IO functions, if you keep reading the async docs, but if none of your IO functions are coroutines, you should consider not using async for your component and hass will do all the heavy lifting.

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+    def async_update_availability(self):
+        self._available = self.cec_client.lib_cec.PollDevice(self._logical_address)
+        if not self._available:
+            self.remove()
```
There is no need to remove it if you write it to the event bus right away again.

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+
+    @asyncio.coroutine
+    def async_request_physical_address(self):
+        self.cec_client.tx(
```
No I/O inside coroutines.

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+
+    @asyncio.coroutine
+    def async_request_cec_vendor(self):
+        self.cec_client.tx(
```
No I/O inside coroutines.

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+
+    @asyncio.coroutine
+    def async_request_cec_osd_name(self):
+        self.cec_client.tx(
```
No I/O inside coroutines.

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+
+    @asyncio.coroutine
+    def async_request_cec_power_status(self):
+        self.cec_client.tx(
```
No I/O inside coroutines.

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+        DEVICE_PRESENCE[logical] = True
+        _LOGGER.info("Initializing CEC device %s", self.name)
+        self.cec_client = cec_client
+        hass.bus.listen(EVENT_CEC_COMMAND_RECEIVED, self._update_callback)
```
Your component and platforms should not communicate via the event bus. The event bus is for entity related activities. Instead, create a class that each entity can register with and that distributes the data as it comes in.

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+            DEVICE_TYPE_NAMES[self._cec_type_id], self._logical_address, self._name)
+
+    @property
+    def state(self):
```
You cannot overwrite the state as this is a generic entity. State property is owned by the abstract base classes of the components (like media player)

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+        return icon_by_type(self._cec_type_id) if self._icon is None else self._icon
+
+    @property
+    def hidden(self):
```
Don't hide it if it is not available . Hidden should only be used for system entities (like auto created groups)

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -4,118 +4,535 @@
+        return self._state
+
+    @property
+    def device_state_attributes(self):
```
You have this twice.

balloob commented on homeassistant/components/media_player/hdmi_cec.py:
```diff
@@ -0,0 +1,64 @@
+        device in CEC_DEVICES['media_player'])
+
+
+class CecMediaPlayer(CecDevice, MediaPlayerDevice):
```
Why create this as a media player if it is not implementing anything that a media player needs?

konikvranik replied:
This was just preparation for implementing volume up, volume down and other stuf. But removing from master until finished.

balloob commented on homeassistant/components/switch/hdmi_cec.py:
```diff
@@ -0,0 +1,58 @@
+    if ATTR_NEW in discovery_info:
+        _LOGGER.info("Setting up devices %s", discovery_info[ATTR_NEW])
+        add_devices(CecSwitch(hass, CEC_CLIENT, device) for device in
+                    filter(lambda x: x not in DEVICE_PRESENCE or not DEVICE_PRESENCE[x], discovery_info[ATTR_NEW]))
```
Please don't use `filter`. It does not give readable code.

balloob commented on homeassistant/components/switch/hdmi_cec.py:
```diff
@@ -0,0 +1,58 @@
+
+    def turn_on(self, **kwargs) -> None:
+        """Turn device on."""
+        _LOGGER.info("***************** turn_on *****************")
```
Please remove these logging

balloob commented on homeassistant/components/switch/hdmi_cec.py:
```diff
@@ -0,0 +1,58 @@
+        _LOGGER.info("***************** turn_on *****************")
+        self.cec_client.lib_cec.PowerOnDevices(self._logical_address)
+        self._state = STATE_ON
+        self.async_update()
```
You can't call `async_update` from outside the event loop

balloob commented on homeassistant/components/switch/hdmi_cec.py:
```diff
@@ -0,0 +1,58 @@
+
+    def turn_off(self, **kwargs) -> None:
+        """Turn device off."""
+        _LOGGER.info("***************** turn_off *****************")
```
Remove comment

balloob commented on homeassistant/components/switch/hdmi_cec.py:
```diff
@@ -0,0 +1,58 @@
+        _LOGGER.info("***************** turn_off *****************")
+        self.cec_client.lib_cec.StandbyDevices(self._logical_address)
+        self._state = STATE_STANDBY
+        self.async_update()
```
No async_update from outside event loop

balloob commented on homeassistant/components/switch/hdmi_cec.py:
```diff
@@ -0,0 +1,58 @@
+        self.async_update()
+        self.schedule_update_ha_state()
+
+    def __init__(self, hass, cec_client, logical):
```
Please put init methods on the top of a class definition.

balloob reviewed (commented):
A lot of changes needed. Biggest one is that all the CEC protocol/vendor specific information will have to be extracted into a third party lib.

Commit:
add remove as job

Commit:
closing cec, no service schemas

Commit:
correct iterate over dictionary

konikvranik committed:
Volume by commands

konikvranik committed:
threading

konikvranik committed:
logging minimized

konikvranik committed:
get load out of main thread

konikvranik committed:
naming cleanup

konikvranik committed:
get load out of main thread

konikvranik committed:
optimized discovery

Commit:
async where possible

Commit:
cleanup logging, constructors first

Commit:
pydoc

Commit:
formatting

Commit:
no async_update from out of loop
no hiding entities
removed redundant device_state_attributes
async updating presence

Commit:
no async

konikvranik committed:
working async cec

Commit:
cec in thirdparty lib

Commit:
cec initialized oudsice

Commit:
working without SIGSEGV

Commit:
rollbacked file changed by mistake

Commit:
sending of commands

Commit:
working with ha

Commit:
using hass loop and device driven updates

Commit:
version up

Commit:
version up

Commit:
Command types in pycec, cleanup for HA integration

Commit:
Removed media player, state moved to switch

Commit:
requirements: pyCEC

Commit:
line width to 79

Commit:
doc

Commit:
doc

konikvranik commented:
As you asked, I moved CEC specific stuff into separate library https://github.com/konikvranik/pycec

Commit:
overindentation solved

Commit:
HDMI to uppercase

Commit:
minimal dependency on cec

Commit:
removed unwanted line

Commit:
doc wording

Commit:
margin 79

Commit:
line continuation indent

Commit:
imperative doc

Commit:
lint: indentation

Commit:
fixed overindented

Commit:
fixed overindented

Commit:
fixed overindented

Commit:
fixed overindented

Commit:
order of imports

Commit:
PEP8

konikvranik committed:
keep signature of overriding

konikvranik committed:
removed redundant blank line

rytilahti commented:
Thanks a lot for your work, I'm eagerly waiting for this to get merged into dev. It'd be nice if you'd merge related commits (even for this published PR, if it's okay for HASS developers) or at least for the future ones. :)

konikvranik commented:
@rytilahti Thank you. I created PR to master. So should I create PR into dev branch?

konikvranik committed:
fixed update call method (#4)

kellerza commented:
We can squash before we merge, but you should really be running local tests! Try the shortcut:

```
$ script/lint --changed
```

If you add new files without tests, you should add them to `coveragerc`

kellerza commented:
The PR is against `homeassistant:dev` so no worries. You can have it in any branch in your repo (but makes your life easier if you work in a new branch in your repo)

konikvranik committed:
Preparation for merge to upstream (#5)

* newer version of pyCEC
* updated services.yaml
* fixed lint scrpt to operate only on python files

* pycec version up

* update services

* no coverage report

* exclude non python files from lint

* lint only on python files

konikvranik commented:
@kellerza So I added also switch/hdmi_cec into .coveragerc.

    $ script/lint --changed

Now passing without error, but I have had to modify it to check python files only. It failed on "syntax error"  in services.yaml.

I also run whole tox lint and also passed.

Only I suspect to miss now is to slightly update documentation on home-assistant.github.io

kellerza commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -1,121 +1,290 @@
+        hass.services.register(DOMAIN, SERVICE_VOLUME, _volume,
+                               descriptions[SERVICE_VOLUME])
+        hass.services.register(DOMAIN, SERVICE_UPDATE_DEVICES, _update,
+                               descriptions[SERVICE_UPDATE_DEVICES])
```
You should add the schemas you defined when registering these services

kellerza commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -1,121 +1,290 @@
+        self._state = STATE_UNKNOWN
+        self._logical_address = logical
+        self.entity_id = "%s.%d" % (DOMAIN, self._logical_address)
+        device.set_update_callback(self._update)
```
This also indicates the data you sent over the buss is not serializable.

Rather send non-serializable data in the `hass.data` dictionary.

kellerza commented on script/lint:
```diff
@@ -4,7 +4,7 @@
 
 if [ "$1" = "--changed" ]; then
-  export files="`git diff upstream/dev --name-only | grep -v requirements_all.txt`"
+  export files="`git diff upstream/dev --name-only | grep -e '\.py$'`"
```
👍 I like this, but do you mind doing it in a separate PR?

konikvranik replied:
OK, in #5018

kellerza commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -1,121 +1,290 @@
+    def _volume(call):
+        """Increase/decrease volume and mute/unmute system."""
+        for cmd, att in call.data.items():
+            att = int(att)
```
Once you have the SCHEMAs validating your input, this type of coercing variables should not be needed

kellerza reviewed (changes requested):
I have not looked at your custom library, but you should make sure that the component you created can function (and most importantly that you can test it!) outside of HASS

konikvranik commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -1,121 +1,290 @@
+    def _new_device(device):
+        """Called when new device is detected by HDMI network."""
+        discovery.load_platform(hass, "switch", DOMAIN,
+                                discovered={ATTR_NEW: [device]},
```
So should I put device into hasss.data with my unique key and then retrieve it by this key which will be send over load_platform?

kellerza replied:
Yes, you can make your `DOMAIN` part of the key, then you know it should be unique

konikvranik committed:
Dev (#6)

* reordered
* sending nonserialized data through hass.data
* code formatting
* code formatting
* import order

konikvranik committed:
Dev (#7)

* newer version of pyCEC
* updated services.yaml
* fixed lint scrpt to operate only on python files

* pycec version up

* update services

* no coverage report

* exclude non python files from lint

* lint only on python files

* reordered

* sending nonserialized data through hass.data

* import order

* fixed object handling

* code formatting

konikvranik commented:
@kellerza
Yes, pyCEC is able to run alone and I tested it on RPi (only platform I have where libcec works with python3)

I also tested it together with Home Assistant on RPi and it works.

Furthermore I work on improving test coverage of pyCEC.

happyleavesaoc commented:
Does this PR retain existing CEC functionality? Specifically the ability to switch to a non-CEC-enabled source?

konikvranik commented:
@happyleavesaoc 
What does mean "switch to a non-CEC-enabled source"?

Code in PR allows you to send arbitraty CEC command, so you can send "standby" or "keypress power on/off" or "set active source" for example. See [cec-o-matic](http://www.cec-o-matic.com/).

Devices are turned on/off over it's entites.

happyleavesaoc commented:
@konikvranik https://home-assistant.io/components/hdmi_cec/#select-device

I know you can send an arbitrary CEC command to achieve that, but the current `cec` component will construct that for you based on the configuration file.

Btw, for context, I'm the author of the current `cec` component. I do like what you're doing here - just want to make sure we don't lose anything in terms of usability.

konikvranik commented:
@happyleavesaoc
I did not implement previous services yet, but I'm planning to do it as soon as possible. See https://github.com/konikvranik/pyCEC/issues/7

I'm not sure about matrix in config. As cec natively works with logical addresses 0x0 - 0xf.
But it is possible to override devices name in customization as switch.hdmi_1: Recorder.
But addressing by friendly_names could be a trap. User can define duplicities. So I prefer to address devices by entity id which is switch.hdmi_? for now. Is it applicable to you? In this case you will have to change your friendly name in automation to entity id.

And thank you for this component and great tutorial to make cec working! :-) I really appreciate it!

konikvranik committed:
Backwards compatibility of hdmi_cec (#10)

* services:
power_on
standby
active_source

konikvranik committed:
new version of pyCEC (#12)

* newer version of pyCEC

konikvranik committed:
devices config (#13)

* getting device name from config

konikvranik commented:
@happyleavesaoc 
Reimplemented previous services power_on, standby and select_device.

The only change is format in configuration. It is single level dictionary "device name":"physical address".
When you call select_device, you can use direct physical address or entity_id or alias preconfigured in config.

konikvranik committed:
shutdown fix (#14)


* correct call on shutdown

konikvranik committed:
remove misplaced annotations (#15)

* Preparation for merge to upstream (#5)

* newer version of pyCEC
* updated services.yaml
* reordered
* sending nonserialized data through hass.data
* services:
power_on
standby
active_source
* code formatting
* getting device name from config
* correct call on shutdown

konikvranik committed:
pyCEC version 0.3.6 (#18)

* newer version of pyCEC
* updated services.yaml
* sending nonserialized data through hass.data
* services:
** power_on
** standby
** active_source
* getting device name from config
* correct call on shutdown
* fork new thread on multicore machines
* support both config schemas: original and new (#16)
* volume press and release support (#17)

happyleavesaoc commented:
@konikvranik Great! Thanks for maintaining that feature. I'm happy to see someone really expanding CEC.

konikvranik committed:
support for media_player (#21)

* accept hexadecimal format of commands
* support for media player
* platform customization
* type constants

konikvranik committed:
Dev (#23)

* accept hexadecimal format of commands
* support for media player
* platform customization

konikvranik committed:
TCP CEC support (#24)

* accept hexadecimal format of commands
* support for media player
* platform customization
* preparing tcp support

konikvranik committed:
volume handling (#25)

kellerza commented on homeassistant/components/test.json:
```diff
@@ -0,0 +1,32 @@
+{
+  'homeassistant': OrderedDict
```
I see github colours these red. Proper JSON delimiters probably only include double quotes `"`. It's only Python that is very lenient

konikvranik replied:
This file was added by mistake. I'm removing it.

kellerza commented on homeassistant/components/services.yaml:
```diff
@@ -153,3 +153,58 @@ verisure:
+
+    fields:
+      up:
+        description: Increases volume x times.
```
x **steps** or x **levels** might be a better term

kellerza commented on homeassistant/components/media_player/hdmi_cec.py:
```diff
@@ -0,0 +1,181 @@
+    def _update(self, device=None):
+        """Update device status."""
+        if device:
+            from pycec.const import STATUS_PLAY
```
Please group these

kellerza commented on homeassistant/components/media_player/hdmi_cec.py:
```diff
@@ -0,0 +1,181 @@
+
+    def send_keypress(self, key):
+        """Send keypress to CEC adapter."""
+        from pycec.commands import KeyPressCommand
```
Please group these

kellerza commented on homeassistant/components/services.yaml:
```diff
@@ -153,3 +153,58 @@ verisure:
+    description: Update devices state from network.
+
+  volume:
+    description: Increase or decrease volume of system
```
Please try to be consistent with either having or omitting `.` across your file.

kellerza commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -30,92 +113,307 @@
+    def _update(self, device=None):
+        """Update device status."""
+        if device:
+            from pycec.const import STATUS_PLAY
```
Please group these imports. If you add a newline immediately after opening a bracket you have lots of space for the constant names, i.e.:

```python
from pycec.const import (
    STATUS_PLAY, STATUS_STOP, POWER_OFF, POWER_ON)
```

kellerza reviewed (approved):
@konikvranik this is looking good! Some minor style comments and should be ok to go.

@happyleavesaoc can you test your existing use case?

konikvranik committed:
Incorporated CR remarks (#26)

* cleanup imports
* cleanup and enhance services description
* removed unwanted file

kellerza commented:
Everything looks good to me, thanks @konikvranik! 🐬 

Let's give @happyleavesaoc a couple more days to test, else we can merge as is

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -1,27 +1,110 @@
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity import Entity
+
+REQUIREMENTS = ['pyCEC>=0.4.5']
```
please pin dependency versions

balloob commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -1,27 +1,110 @@
+    vol.Optional(ATTR_DST): _VOL_HEX,
+    vol.Optional(ATTR_ATT): _VOL_HEX,
+    vol.Optional(ATTR_RAW): vol.Coerce(str)
+}, extra=vol.ALLOW_EXTRA)
```
services should not allow extra data.

konikvranik committed:
implemented CR remarks (#27)

* pyCEC v0.4.6
* pined dependency version
* tighten service schemas

happyleavesaoc commented:
I won't have time to test in this timeframe. Sorry!

happyleavesaoc commented:
I did get a chance to test this today. The functions that are supported by my hardware worked. I had one issue that I can't seem to reproduce, at one point it tried to add an entity_id (switch.hdmi_x) that already existed.

happyleavesaoc commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -1,27 +1,110 @@
+CMD_DOWN = 'down'
+CMD_MUTE = 'mute'
+CMD_UNMUTE = 'unmute'
+CMD_MUTE_TOGGLE = 'toggle mute'
```
Unused

happyleavesaoc commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -1,27 +1,110 @@
+SERVICE_VOLUME_SCHEMA = vol.Schema({
+    vol.Optional(CMD_UP): vol.Any(CMD_PRESS, CMD_RELEASE, vol.Coerce(int)),
+    vol.Optional(CMD_DOWN): vol.Any(CMD_PRESS, CMD_RELEASE, vol.Coerce(int)),
+    vol.Optional(CMD_MUTE): None,
```
Its kind of awkward to send null-value keys in the service.

konikvranik replied:
Will rearrange mute chema. Created issue konikvranik/home-assistant#31

armills commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -30,92 +113,304 @@
+        hass.loop if multiprocessing.cpu_count() < 2 else None)
+    host = base_config[DOMAIN].get(CONF_HOST, None)
+    if host:
+        adapter = TcpAdapter(host, name="HASS", activate_source=False)
```
Small suggestion: Can we use the full name "Home Assistant" rather than the abbreviation "HASS"?

konikvranik replied:
Will do in konikvranik/home-assistant#30

happyleavesaoc commented on homeassistant/components/hdmi_cec.py:
```diff
@@ -30,92 +113,304 @@
+            hdmi_network.send_command(KeyReleaseCommand(dst=ADDR_AUDIOSYSTEM))
+        else:
+            att = 1 if att == "" else int(att)
+            for _ in range(1, att):
```
Should be `range(0, att)`. Right now, it won't run if `att` is one.

konikvranik replied:
Opened konikvranik/home-assistant#29

rytilahti commented:
Looks like pycec is missing from requirements_all.txt.

happyleavesaoc commented:
One other observation. When using the TCP server `pycec`, and restarting it, Home Assistant loses connection and won't reconnect without a restart itself.

konikvranik commented:
@happyleavesaoc: Opened issue konikvranik/pyCEC#30 to handle client reconnecting.

konikvranik committed:
requirements (#28)

konikvranik commented:
@rytilahti: pycec dependency is present in version 0.4.6. Otherwise checks fail.

kellerza commented:
@konikvranik Thanks for the PR, it took a while to get it in, but you added many features and worth it in the end! 🐬 !!

michaelhenningersrb commented:
Hey could you tell me when this giant feature is available in a new release.... Great... thank you guys

kellerza commented:
It should be the coming weekend
